### PR TITLE
Allow adding attachments in qgz files

### PR DESCRIPTION
**Pull request by m-kuhn**
Allows packaging any kind of files into a qgz file.

There is a new signal `QgsProject::collectAttachedFiles` which allows plugins or core code to attach files.

There are the methods `QgsProject::attachedFile( filename )` and `QgsProject::attachedFiles()` to get access to those files.

Main open question: should attached files be forced into a subdirectory `/resources` or `/attachments` inside the zip to avoid name collisions.

### CODE REVIEW

**Comment by m-kuhn** [comment]
Should this be labelled as a feature? I tend(ed) to not label API changes as features but I noticed that they often are labelled as a feature, so now I'm unsure.

**Comment by nyalldawson** [comment]
I only label API stuff as a feature if I think it's going to be particularly exciting to end user developers (vs core Devs) and is worthy of inclusion in the release notes. In this case, I'd say yes!

**Comment by nyalldawson** [comment]
The current signal based API is good, but I wonder if we also should have an explicit "add file" call which isn't a signal? I'm thinking for code which is short lived and doesn't have a lasting object that can respond to the signal at save time.

**Comment by nyalldawson** [comment]
^^ I meant "also have"

**Comment by m-kuhn** [comment]
> The current signal based API is good, but I wonder if we also should have an explicit "add file" call which isn't a signal? I'm thinking for code which is short lived and doesn't have a lasting object that can respond to the signal at save time.

The current version came out of the need to do something (i.e. thumbnail generation) at save time.
Having a method to explicitly add a file sounds nice, but at the moment this is only implemented for qgz files (and I don't plan to do anything more at the moment) and an explicit call could be interpreted as a hint in this direction.

I added some docs to clarify on these things.

File: src/core/qgsproject.h [hunk 1]
```diff
@@ -980,7 +980,7 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      *
      * \note Attached files are only supported by QGZ file based projects
      * \see collectAttachedFiles()
-     * \see attachedFile( fileName )
+     * \see attachedFile()
```
Comment (nirvn) [inline src/core/qgsproject.h]: @m-kuhn , your sip file is out of sync :)

Comment (m-kuhn) [reply src/core/qgsproject.h]: sip files are never out of sync, they synchronize exactly when they mean to.

**Comment by nirvn** [comment]
@ccrook , your memory layer saver plugin should make use of this new feature, that'd allow you to embed memory layer data straight into the default project file type itself.

**Comment by m-kuhn** [comment]
@nyalldawson I have merged it with your question still open. It's not a strong opinion from my side to not implement an imperative method, but in either case I think it can be done very well as a follow up and shouldn't block this.i

**Comment by m-kuhn** [comment]
@nyalldawson I have merged it with your question still open. It's not a strong opinion from my side to not implement an imperative method, but in either case I think it can be done very well as a follow up and shouldn't block merging this.

**Comment by wonder-sk** [comment]
I have only now managed to look at that... +1 to Nyall's suggestion for explicit calls.

I do not understand the python API docs:
```
 QgsProject.instance().collectAttachedFiles.connect(lambda files: files + ['/absolute/path/to/my/attachment.txt'])
```
How does that work if collectAttachedFiles() works with a string map but the lambda deals with a list?

**Comment by m-kuhn** [comment]
That is absolutely correct, will fix

**Comment by m-kuhn** [comment]
For the explicit call, additional to the above reasons it will add the requirement to keep the file stored somewhere until the project is saved which adds some additional (implementation) overhead which is out of scope for the current project. I'll leave this to anyone else interested in working on this.

**Comment by m-kuhn** [comment]
> How does that work if collectAttachedFiles() works with a string map but the lambda deals with a list?

I checked the python bindings and realized that they are not working at all. I guess sip does not properly implement the`/In,Out/`annotation for signals, but because of the automatic mapping of QgsStringMap to a python dict it's not possible to directly modify the parameter.

In short: the current implementation is broken and doesn't even help plugin devs, so this defeats the whole purpose of it.

If no one has a smart idea, I'll revert this and propose it again as part of another pull request with the `SIP_SKIP` annotation.

**Comment by wonder-sk** [comment]
That's something I was worried about - that this kind of signal-based API for collecting files may have issues due to the fact that QMap gets converted to dict.

I would suggest to have a container class for all the extra files, so it can be passed by non-const reference also to Python. I assume over time we may add some more options/flags to files that get attached, so it would also future proof the API. ...or even better, to have API not based on signal to collect files ;-)

My vision was that QgsReadWriteContext class would be used also for collecting/accessing extra files, so this mechanism could ideally be independent from QgsProject (e.g. when saving a layer style to .qml, we would be able to detect any SVG symbols that are not part of the installation and attach them).

**Comment by m-kuhn** [comment]
That was the kind of approaches I had in mind too, but which are not feasible in the scope of the current project.

Revert pending https://github.com/qgis/QGIS/pull/10019/files

**Comment by m-kuhn** [comment]
> or even better, to have API not based on signal to collect files ;-)

nothing against an API which is not based on signals, but for the purpose I currently have it's just fine because the file should be generated on save.

**Pull request merged**